src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Analyzer.cs

```diff
@@ -0,0 +1,205 @@
+        internal const string StringCompareToMethodName = "CompareTo";
+
+        private static readonly LocalizableString s_localizableStringComparisonTitle = CreateLocalizableResourceString(nameof(RecommendCaseInsensitiveStringComparisonTitle));
+        private static readonly LocalizableString s_localizableStringComparisonMessage = CreateLocalizableResourceString(nameof(RecommendCaseInsensitiveStringComparisonMessage));
+        private static readonly LocalizableString s_localizableStringComparisonDescription = CreateLocalizableResourceString(nameof(RecommendCaseInsensitiveStringComparisonDescription));
+
+        private static readonly LocalizableString s_localizableStringComparerTitle = CreateLocalizableResourceString(nameof(RecommendCaseInsensitiveStringComparerTitle));
+        private static readonly LocalizableString s_localizableStringComparerMessage = CreateLocalizableResourceString(nameof(RecommendCaseInsensitiveStringComparerMessage));
+        private static readonly LocalizableString s_localizableStringComparerDescription = CreateLocalizableResourceString(nameof(RecommendCaseInsensitiveStringComparerDescription));
+
+        internal static readonly DiagnosticDescriptor RecommendCaseInsensitiveStringComparisonRule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            s_localizableStringComparisonTitle,
+            s_localizableStringComparisonMessage,
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            s_localizableStringComparisonDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        internal static readonly DiagnosticDescriptor RecommendCaseInsensitiveStringComparerRule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            s_localizableStringComparerTitle,
+            s_localizableStringComparerMessage,
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            s_localizableStringComparerDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(
+            RecommendCaseInsensitiveStringComparisonRule, RecommendCaseInsensitiveStringComparerRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(AnalyzeCompilationStart);
+        }
+
+        private void AnalyzeCompilationStart(CompilationStartAnalysisContext context)
+        {
+            // Retrieve the essential types: string, StringComparison, StringComparer
+
+            if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemString, out INamedTypeSymbol? stringType))
```

Comment:
I'd do it as `context.Compilation.GetSpecialType(SpecialType.System_String)`. It's never null, but *could* be error symbol.

Reply:
> but _could_ be error symbol

Interesting. Do you have any idea why?

Reply:
> context.Compilation.GetSpecialType(SpecialType.System_String)

If this is preferred for special types, we could create an analyzer to suggest that 😄

Reply:
Suggestion applied.